test-interface-common/src/main/scala/scala/scalanative/testinterface/common/RPCCore.scala

```diff
@@ -0,0 +1,297 @@
+            val throwable = Try(deserialize[Throwable](in)) match {
+              case Success(t) => new RPCException(t)
+              case Failure(t) => t
+            }
+            p.promise.failure(throwable)
+          }
+
+        case _ =>
+          endpoints.get(opCode) match {
+            case null =>
+              /** Quick and dirty way to provide more error detail for certain
+               * known problems.
+               * This is not ideal, but the best we can do, since we do not know
+               * all possible opCodes we could receive (we'd need something like
+               * an opCode "domain").
+               * For now this is good enough; if collisions happen in the
+               * future, we can improve this.
+               */
+              val detail = opCode match {
+                case NativeEndpoints.msgWorker.opCode =>
+                  "; " +
+                    "The test adapter could not send a message to a worker, " +
+                    "which probably happens because the worker terminated early, " +
+                    "without waiting for the reply to a call to send(). " +
+                    "This is probably a bug in the testing framework you are " +
+                    "using. See also scala-js/scala-js#3201."
+
+                case _ => ""
+              }
+
+              throw new IllegalStateException(s"Unknown opcode: $opCode$detail")
+
+            case bep: BoundMsgEndpoint =>
+              val ep: bep.endpoint.type = bep.endpoint
+              import ep._
+              val arg = deserialize[Msg](in)
+              bep.exec(arg)
+
+            case bep: BoundRPCEndpoint =>
+              val callID = in.readLong()
+
+              val ep: bep.endpoint.type = bep.endpoint
+              import ep._
+
+              Future
+                .fromTry(Try(deserialize[Req](in)))
+                .flatMap(bep.exec)
+                .onComplete(repl => send(makeReply(callID, repl)))
+          }
+
+      }
+    }
+  }
+
+  /** Subclass needs to implement message sending. */
+  protected def send(msg: String): Unit
+
+  /** Used to send a message to the other end. */
+  final def send(ep: MsgEndpoint)(msg: ep.Msg): Unit = {
+    import ep._
+    send(makeMsgMsg(opCode, msg))
+  }
+
+  /** Used to make an actual call to the other end. */
+  final def call(ep: RPCEndpoint)(req: ep.Req): Future[ep.Resp] = {
+    import ep._
+
+    // Reserve an id for this call.
+    val id = nextID.incrementAndGet()
+
+    // Prepare message. We do this early in case it throws.
+    val msg = makeRPCMsg(opCode, id, req)
+
+    // Register pending call.
+    val promise = Promise[Resp]()
+    val oldCall = pending.put(id, PendingCall(promise))
+
+    if (oldCall != null) {
+      val error = new AssertionError("Ran out of call ids!")
+      close(error)
+      throw error
+    }
+
+    if (closeReason != null) {
+
+      /** In the meantime, someone closed the channel. Help closing.
```

Comment:
```suggestion
      /* In the meantime, someone closed the channel. Help closing.
```
(this is not a Scaladoc comment)